File: src/main/java/com/almasb/zeph/combat/Attribute.java
package com.almasb.zeph.combat;

/**
 * 9 primary attributes of a game character
 *
 * @author Almas Baimagambetov (ab607@uni.brighton.ac.uk)
 * @version 1.0
 *
 */
public enum Attribute {
    STRENGTH("Increases damage dealt by physical attacks. Also slightly increases HP and regeneration."),
    VITALITY("Reduces damage you take, increases maximum HP and regeneration."),
    DEXTERITY("Increases attack speed and slightly improves all other combat stats."),
    AGILITY("Allows you to attack a lot faster with all weapons."),
    INTELLECT("Greatly increases magic damage and defense. Also provides greater SP pool."),
    WISDOM("Significantly increases SP pool and regeneration. Slightly increases magic damage."),
    WILLPOWER("Allows you to withstand more magic damage and score critical hits with skills."),
    PERCEPTION("Slightly improves all stats, including chance of critical hits with attacks and skills."),
    LUCK("Directly affects the ability to score critical hits and deal more critical damage.");

    private final String description;

    private Attribute(String description) {
        this.description = description;
    }

    /**
     *
     * @return attribute description
     */
    public String getDescription() {
        return description;
    }

    @Override
    public String toString() {
        return this.name().length() > 3 ? this.name().substring(0, 3) : this.name();
    }

    /**
     * Data structure for an info object that contains
     * values of all the attributes
     */
    public static class AttributeInfo {

        public int str = 1, vit = 1, dex = 1,
                agi = 1, int_ = 1, wis = 1,
                wil = 1, per = 1, luc = 1;

        public AttributeInfo str(final int value) {
            str = value;
            return this;
        }

        public AttributeInfo vit(final int value) {
            vit = value;
            return this;
        }

        public AttributeInfo dex(final int value) {
            dex = value;
            return this;
        }

        public AttributeInfo agi(final int value) {
            agi = value;
            return this;
        }

        public AttributeInfo int_(final int value) {
            int_ = value;
            return this;
        }

        public AttributeInfo wis(final int value) {
            wis = value;
            return this;
        }

        public AttributeInfo wil(final int value) {
            wil = value;
            return this;
        }

        public AttributeInfo per(final int value) {
            per = value;
            return this;
        }

        public AttributeInfo luc(final int value) {
            luc = value;
            return this;
        }
    }
}

File: src/main/kotlin/com/almasb/zeph/combat/Attribute.kt
package com.almasb.zeph.combat

import com.almasb.zeph.combat.Attribute.AM.*
import com.almasb.zeph.combat.Stat.*

/**
 * Primary attributes (9 in total) of a game character.
 *
 * @author Almas Baimagambetov
 */
enum class Attribute(description: String, vararg cStats: Pair<Stat, AM>) {

    STRENGTH("Increases damage dealt by physical attacks. Increases HP and regeneration.",
            ATK increase VERY_HIGH,
            MAX_HP increase MEDIUM,
            HP_REGEN increase LOW),

    VITALITY("Reduces damage you take. Increases HP and regeneration.",
            MAX_HP increase (VERY_HIGH),
            HP_REGEN increase (HIGH),
            DEF increase (HIGH)),

    DEXTERITY("Increases attack speed and improves offense based stats.",
            ATK increase (MEDIUM),
            MATK increase (MEDIUM),
            ASPD increase (LOW),
            MSPD increase (LOW)),

    AGILITY("Allows to attack faster with all weapons.",
            ASPD increase (VERY_HIGH)),

    INTELLECT("Increases magic damage and defense. Provides greater SP pool.",
            MATK increase (VERY_HIGH),
            MDEF increase (MEDIUM),
            MAX_SP increase (LOW)),

    WISDOM("Increases SP pool and regeneration. Increases magic damage.",
            MATK increase (VERY_LOW),
            MDEF increase (VERY_LOW),
            MAX_SP increase (VERY_HIGH),
            SP_REGEN increase (VERY_HIGH)),

    WILLPOWER("Increases magic defense and SP pool. Increases critical hit chance with skills",
            MCRIT_CHANCE increase (HIGH),
            MDEF increase (VERY_HIGH),
            MAX_SP increase (LOW),
            MSPD increase (MEDIUM)),

    PERCEPTION("Improves all combat stats.",
            ATK increase (LOW),
            MATK increase (LOW),
            DEF increase (VERY_LOW),
            MDEF increase (VERY_LOW),
            CRIT_CHANCE increase (VERY_LOW),
            MCRIT_CHANCE increase (VERY_LOW)),

    LUCK("Affects the ability to score critical hits and deal more critical damage.",
            CRIT_CHANCE increase (VERY_HIGH),
            MCRIT_CHANCE increase (VERY_HIGH),
            CRIT_DMG increase (VERY_LOW),
            MCRIT_DMG increase (VERY_LOW));

    val description: String

    @Suppress("UNCHECKED_CAST")
    val stats: Array<Pair<Stat, AM> > = cStats as Array<Pair<Stat, AM> >

    init {
        this.description = "$description\n" + stats.joinToString("\n") { pair -> "%12s %-5s".format(pair.first, pair.second.desc()) }
    }

    override fun toString() = this.name.substring(0, 3)

    /**
     * Attribute modifier.
     */
    enum class AM {
        VERY_LOW, LOW, MEDIUM, HIGH, VERY_HIGH;

        fun value() = (ordinal + 1) * 0.1

        fun desc() = "+".repeat(ordinal + 1)
    }

    /**
     * Data structure for an info object that contains
     * values of all the attributes.
     */
    class AttributeInfo {
        val attrs = hashMapOf<Attribute, Int>()

        init {
            // add 0 defaults
            Attribute.values().forEach {
                it +0
            }
        }

        operator fun Attribute.plus(value: Int) {
            attrs[this] = value
        }
    }
}

private infix fun Stat.increase(am: Attribute.AM) = this to am

/**
 * Increases an attribute.
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class Rune(val attribute: Attribute, val bonus: Int) {

    override fun toString() = "$attribute +$bonus"
}